base/show.jl

```diff
@@ -1432,25 +1504,49 @@ function show_unquoted(io::IO, ex::Expr, indent::Int, prec::Int)
         show_linenumber(io, args...)
 
     elseif head === :try && 3 <= nargs <= 4
-        show_block(io, "try", args[1], indent)
+        show_block(io, "try", args[1], indent, quote_level)
         if is_expr(args[3], :block)
-            show_block(io, "catch", args[2] === false ? Any[] : args[2], args[3], indent)
+            show_block(io, "catch", args[2] === false ? Any[] : args[2], args[3], indent, quote_level)
         end
         if nargs >= 4 && is_expr(args[4], :block)
-            show_block(io, "finally", Any[], args[4], indent)
+            show_block(io, "finally", Any[], args[4], indent, quote_level)
         end
         print(io, "end")
 
     elseif head === :block
-        show_block(io, "begin", ex, indent)
+        show_block(io, "begin", ex, indent, quote_level)
         print(io, "end")
 
     elseif head === :quote && nargs == 1 && isa(args[1], Symbol)
-        show_unquoted_quote_expr(io, args[1]::Symbol, indent, 0)
+        show_unquoted_quote_expr(io, args[1]::Symbol, indent, 0, quote_level+1)
+    elseif head === :quote && nargs == 1 && Meta.isexpr(args[1], :block)
+        if length(args[1].args) == 1
+            # one argument: Expr(:quote, Expr(:block, a_single_arg))
```

Comment:
On the countrary, I see that this branch is needed to correctly handle things like
`macro m(a, b)
    quote$a + $b
    end
end`

Reply:
On the countrary, tests suggest that this branch is needed to correctly handle cases like
```
macro m(a, b)
    quote
        \$a + \$b
    end
end
```
In fact, without the branch
```
julia> repr(Meta.parse("""
       macro m(a, b)
           quote
               \$a + \$b
           end
       end"""))
":(macro m(a, b)\n      #= none:2 =#\n      :(begin\n            #= none:3 =#\n            \$a + \$b\n        end)\n  end)"
```
while with it in place, more satisfactorily the result is
```
julia> repr(Meta.parse("""
       macro m(a, b)
           quote
               \$a + \$b
           end
       end"""))
":(macro m(a, b)\n      #= none:2 =#\n      quote\n          #= none:3 =#\n          \$a + \$b\n      end\n  end)"
```

Another case that fails without the branch is
```
quote
    quote
        \$\$x
    end
end
```
with both quotes turned into `:(begin ... end)` upon `repr(Meta.parse("..."))`ing it:
```
julia> repr(Meta.parse("""
       quote
           quote
               \$\$x
           end
       end
       """))
":(:(begin\n        #= none:2 =#\n        :(begin\n              #= none:3 =#\n              \$\$x\n          end)\n    end))"
```

Reply:
In fact I've checked now, and actually `"quote \$x end"` is parsed as a `quote Expr` containing a `block Expr`:
```
julia> dump(Meta.parse("quote \$x end"))
Expr
  head: Symbol quote
  args: Array{Any}((1,))
    1: Expr
      head: Symbol block
      args: Array{Any}((2,))
        1: LineNumberNode
          line: Int64 1
          file: Symbol none
        2: Expr
          head: Symbol $
          args: Array{Any}((1,))
            1: Symbol x
```

Reply:
Yes, those cases are fine. The problem is when the `block` Expr has one element, e.g. it prints the value of `Expr(:quote, Expr(:block, :(a+b)))` as `:(:(a + b))`.

Reply:
After some thinking, I must say that I very much prefer having
```
julia> show(Expr(:quote, Expr(:block, :(a + b))))
:(:(a + b))
```
than
```
julia> show(Expr(:quote, Expr(:block, :(a + b))))
:(:(begin
        a + b
    end))
```
since the former is coherent with the case in which the block has more than one argument, where, for reasons mentioned in my last two comments, the block itself is actually hidden upon `show`ing:
```
julia> show(Expr(:quote, Expr(:block, :(a + b), :(c + d))))
:(quote
      a + b
      c + d
  end)
```
Nevertheless, the final word is of course yours.

Reply:
Well, we shouldn't print two ASTs the same way if they're actually different trees. The parser doesn't even emit a block for the syntax `:(a+b)`:

```julia
julia> dump(:(:(a+b)))
Expr
  head: Symbol quote
  args: Array{Any}((1,))
    1: Expr
      head: Symbol call
      args: Array{Any}((3,))
        1: Symbol +
        2: Symbol a
        3: Symbol b

julia> dump(Expr(:quote, Expr(:block, :(a+b))))
Expr
  head: Symbol quote
  args: Array{Any}((1,))
    1: Expr
      head: Symbol block
      args: Array{Any}((1,))
        1: Expr
          head: Symbol call
          args: Array{Any}((3,))
            1: Symbol +
            2: Symbol a
            3: Symbol b
```

Reply:
Yes, exactly. I don't think we can just drop a `:block` expression; for example people often expect them to be there in order to `push!` statements in. Wouldn't it be best to print it as
```
:(quote
      a + b
  end)
```
? I know there's a difference of whether a line number is present, but I don't think that's as important as keeping the extra block nesting.

Reply:
Sounds good to me. Another option could be `:(:(a + b;))` which parses as a block with a single entry and no line number node. What I'm not entirely sure about is whether that works in all cases?

Reply:
Now I see your point. After some experimenting I think my preference is for @JeffBezanson 's suggestion. With that, we get a nicely consistent picture:
```
julia> show(Expr(:quote, Expr(:block)))
:(quote
  end)

julia> show(Expr(:quote, Expr(:block, :(a + b))))
:(quote
      a + b
  end)

julia> show(Expr(:quote, Expr(:block, :(a + b), :(c + d))))
:(quote
      a + b
      c + d
  end)
```
@c42f  's suggestion is somewhat cleaner, as there are no line numbers appearing in the call `Meta.parse(repr(Expr(:quote, Expr(:block, :(a + b)))))`
but I find `:(:(a + b;))` and in particular `:(:(a;))` to be a little confusing, and I definitely dislike it if instead of `a + b` or `a` there is some more complicated expression, possibly spread over more than one line: compare
```
julia> show(Expr(:quote, Expr(:block, Expr(:if, :(a == b), :(a + b), :(a - b)))))
:(:(if a == b
        a + b
    else
        a - b
    end;))
```
with
```
julia> show(Expr(:quote, Expr(:block, Expr(:if, :(a == b), :(a + b), :(a - b)))))
:(quote
      if a == b
          a + b
      else
          a - b
      end
  end)
```
By the way, this is my opinion, and I must admit that I may dislike semicolons more than they possibly deserve.

Reply:
I have updated the test_repr macro, adding a stronger form of the previous condition which have to be satisfied for the test to pass. The current solution, which neglect the one-argument inner block, fails the stronger test, while both of the suggestions due to @JeffBezanson and @c42f pass it. Now it is only a matter of deciding which solution to endorse.

Reply:
Once this is settled, with a cleaner local history I will merge with the upstream and fix conflicts.

Reply:
Let's use the `quote ... end` syntax as you prefer. I don't think there is any particular promise that `:(x;)` will not include a line number; in fact many people would probably prefer if it did include one.